test/unit/BNum.t.sol

```diff
@@ -0,0 +1,342 @@
+    vm.expectRevert(BNum.BNum_SubUnderflow.selector);
+
+    bNum.call_bsub(_a, _b);
+  }
+
+  function test_BsubWhenPassingKnownValues() external {
+    // it should return correct value
+    //     5 - 4.01 = 0.99
+    uint256 _a = 5e18;
+    uint256 _b = 4.01e18;
+
+    uint256 _result = bNum.call_bsub(_a, _b);
+
+    assertEq(_result, 0.99e18);
+  }
+
+  function test_BsubSignWhenPassingZeroAndZero() external {
+    (uint256 _result, bool _flag) = bNum.call_bsubSign(0, 0);
+
+    // it should return zero and false
+    assertEq(_result, 0);
+    assertFalse(_flag);
+  }
+
+  function test_BsubSignWhenPassingALessThanB(uint256 _a, uint256 _b) external {
+    _a = bound(_a, 0, type(uint256).max - 1);
+    _b = bound(_b, _a + 1, type(uint256).max);
+
+    (uint256 _result, bool _flag) = bNum.call_bsubSign(_a, _b);
+
+    // it should return correct value and true
+    assertEq(_result, _b - _a);
+    assertTrue(_flag);
+  }
+
+  function test_BsubSignWhenPassingKnownValues() external {
+    (uint256 _result, bool _flag) = bNum.call_bsubSign(5e18, 3e18);
+
+    // it should return correct value
+    assertEq(_result, 2e18);
+    assertFalse(_flag);
+  }
+
+  function test_BmulWhenPassingZeroAndZero() external {
+    uint256 _result = bNum.call_bmul(0, 0);
+
+    // it should return zero
+    assertEq(_result, 0);
+  }
+
+  function test_BmulRevertWhen_PassingAAsUint256MaxAndBNonZero(uint256 _b) external {
+    uint256 _a = type(uint256).max;
+    _b = bound(_b, 1, type(uint256).max);
+
+    // it should revert
+    vm.expectRevert(BNum.BNum_MulOverflow.selector);
+
+    bNum.call_bmul(_a, _b);
+  }
+
+  function test_BmulRevertWhen_PassingBAsUint256MaxAndANonZero(uint256 _a) external {
+    _a = bound(_a, 1, type(uint256).max);
+    uint256 _b = type(uint256).max;
+
+    // it should revert
+    vm.expectRevert(BNum.BNum_MulOverflow.selector);
+
+    bNum.call_bmul(_a, _b);
+  }
+
+  function test_BmulWhenPassingAMulBTooBig(uint256 _a, uint256 _b) external {
+    _a = bound(_a, type(uint256).max / 2, type(uint256).max);
+    _b = bound(_b, type(uint256).max / 2, type(uint256).max);
+
+    // it should revert
+    //     a * b + BONE / 2 > uint256 max
+    vm.expectRevert(BNum.BNum_MulOverflow.selector);
+
+    bNum.call_bmul(_a, _b);
+  }
+
+  function test_BmulWhenPassingKnownValues() external {
+    // it should return correct value
+    //     1.25 * 4.75 = 5.9375
+
+    uint256 _a = 1.25e18;
+    uint256 _b = 4.75e18;
+
+    uint256 _result = bNum.call_bmul(_a, _b);
+
+    assertEq(_result, 5.9375e18);
+  }
+
+  function test_BdivRevertWhen_PassingBAsZero() external {
+    // it should revert
+    vm.expectRevert(BNum.BNum_DivZero.selector);
+
+    bNum.call_bdiv(1, 0);
+  }
+
+  function test_BdivRevertWhen_PassingAAsUint256Max() external {
+    // it should revert
+    vm.expectRevert(BNum.BNum_DivInternal.selector);
+
+    bNum.call_bdiv(type(uint256).max, 1);
+  }
+
+  function test_BdivWhenPassingAAndBTooBig() external {
```

Comment:
this doesn't implement A and B 'too big', A is max (definetely too big) but b is just BONE

Reply:
corrected it as there are 2 errors in BNum that we wanna reach, the 1st one is when `a * BONE` overflows (soon to be renamed to left / right?), and the 2nd one is the one we wanna catch here, that has a `b` crippled (bc is `/2`) and the sum needs to overflow, interesting puzzle